pkg/controller/certificatemanager/certificatemanager.go

```diff
@@ -207,6 +207,51 @@ func (cm *certificateManager) GetOrCreateKeyPair(cli client.Client, secretName,
 	}, nil
 }
 
+// This type will be returned for errors that do not have the correct Ext Key usage types
+// for a specific secert certificate.
+type CertExtKeyUsageError struct {
+	msg string
+}
+
+func (cue *CertExtKeyUsageError) Error() string {
+	return cue.msg
+}
+
+var _ error = &CertExtKeyUsageError{}
+
+// Returns true if the error is a CertExtKeyUsageError
+func IsCertExtKeyUsageError(err error) bool {
+	_, ok := err.(*CertExtKeyUsageError)
+	return ok
+}
+
+func extKeyUsageToString(requiredExtKeyUsages []x509.ExtKeyUsage) string {
+	out := ""
+	for i, x := range requiredExtKeyUsages {
+		if i != 0 {
+			out = out + ", "
+		}
+		switch x {
+		case x509.ExtKeyUsageServerAuth:
+			out = out + "ExtKeyUsageServerAuth"
+		case x509.ExtKeyUsageClientAuth:
+			out = out + "ExtKeyUsageClientAuth"
+		default:
+			out = out + "unknown"
+		}
+
+	}
+	return out
+}
+
+func newCertExtKeyUsageError(name, ns string, requiredExtKeyUsages []x509.ExtKeyUsage) error {
```

Comment:
I should have been more clear - this is fine for now, but I think a better pattern in Go when dealing with custom error types is to have typed fields in the error itself, and then build the error message in the `Error()` method itself, rather than constructing the message in the constructor.

It allows potential callers to actually access relevant details. For example, if calling code wanted to switch on error type and pull out the secret name and namespace, that would be easy to do with `e.Namespace` instead of needing to parse an error string.

We're not doing anything like that in this PR so it's moot, more or just a best-practice thing.

Reply:
Ah I see what you mean.